pkg/processor/trigger/kafka/trigger.go

```diff
@@ -364,11 +279,115 @@ func (k *kafka) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.C
 	close(submittedEventChan)
 	close(explicitAckControlMessageChan)
 	close(workerDrainingCompleteChan)
-	close(readyForRebalanceChan)
 
 	return submitError
 }
 
+func (k *kafka) drainOnRebalance(session sarama.ConsumerGroupSession,
+	claim sarama.ConsumerGroupClaim,
+	workerInstance *worker.Worker,
+	submittedEventInstance *submittedEvent,
+	message *sarama.ConsumerMessage,
+	finishHandlingMsg bool) {
+
+	k.Logger.DebugWith("Got signal to stop consumption",
+		"wait", k.configuration.maxWaitHandlerDuringRebalance.String(),
+		"partition", claim.Partition(),
+		"finishHandlingMsg", finishHandlingMsg,
+	)
+
+	readyForRebalanceChan := make(chan bool)
+	defer close(readyForRebalanceChan)
+
+	// indicate whether this partition worker was drained
+	// this is used to avoid race condition where 2 different partitions sharing the same worker
+	// will both try to reset the drain flag needlessly
+	var drainedWorker bool
+
+	go func() {
+		defer common.CatchAndLogPanicWithOptions(k.ctx, // nolint: errcheck
+			k.Logger,
+			"Recover from panic after trying to write into channel, which was closed because of wait for rebalance timeout",
+			&common.CatchAndLogPanicOptions{
+				Args:          []interface{}{"partition", claim.Partition()},
+				CustomHandler: nil,
+			})
+		var wg sync.WaitGroup
+		if finishHandlingMsg {
+			wg.Add(2)
+			go func() {
+				err := <-submittedEventInstance.done
+
+				// we successfully submitted the message to the handler. mark it
+				if err == nil {
+					session.MarkOffset(
+						message.Topic,
+						message.Partition,
+						message.Offset+1-int64(k.configuration.ackWindowSize),
+						"",
+					)
+				}
+				k.Logger.DebugWith("Handler done", "partition", claim.Partition())
+				wg.Done()
+			}()
+		} else {
+			wg.Add(1)
+		}
+
+		go func() {
+			// this needs to occur once. the reason is that this specific function (ConsumeClaim)
+			// runs in parallel for each partition, and we want to make sure that we only
+			// drain the workers once.
+			if err := k.SignalWorkerDraining(); err != nil {
+				k.Logger.DebugWith("Failed to signal worker draining",
+					"err", err.Error(),
+					"partition", claim.Partition())
+			} else {
+				drainedWorker = true
+			}
+			wg.Done()
+		}()
+
+		wg.Wait()
+		readyForRebalanceChan <- true
+	}()
+
+	// wait a for rebalance readiness or max timeout
+	select {
+	case <-readyForRebalanceChan:
+		k.Logger.DebugWith("Handler done, rebalancing will commence",
+			"partition", claim.Partition())
```

Comment:
This log should be after the wait below, because rebalance does not necessarily commence now.